File: src/main/java/com/equalinformation/poc/akka/java/track1/wordcount/actors/ReduceActor.java
package com.equalinformation.poc.akka.java.track1.wordcount.actors;

import akka.actor.UntypedActor;
import com.equalinformation.poc.akka.java.track1.wordcount.messages.MapData;
import com.equalinformation.poc.akka.java.track1.wordcount.messages.ReduceData;
import com.equalinformation.poc.akka.java.track1.wordcount.messages.WordCount;

import java.util.HashMap;
import java.util.List;

/**
 * Created by bpupadhyaya on 2/10/16.
 */
public class ReduceActor extends UntypedActor {

    @Override
    public void onReceive(Object message) throws Exception {
        if (message instanceof MapData) {
            MapData mapData = (MapData) message;
            // reduce the incoming data and forward the result to Master actor
//            getSender().tell(reduce(mapData.getDataList())); //TODO migrate code to akka version in sbt
        } else
            unhandled(message);
    }

    private ReduceData reduce(List<WordCount> dataList) {
        HashMap<String, Integer> reducedMap = new HashMap<String, Integer>();
        for (WordCount wordCount : dataList) {
            if (reducedMap.containsKey(wordCount.getWord())) {
                Integer value = (Integer) reducedMap.get(wordCount.getWord());
                value++;
                reducedMap.put(wordCount.getWord(), value);
            } else {
                reducedMap.put(wordCount.getWord(), Integer.valueOf(1));
            }
        }
        return new ReduceData(reducedMap);
    }
}

File: src/main/scala/com/equalinformation/poc/akka/scala/track1/wordcount/actors/ReduceActor.scala
package com.equalinformation.poc.akka.scala.track1.wordcount.actors

import akka.actor.Actor

import com.equalinformation.poc.akka.scala.track1.wordcount._

/**
  * Created by bpupadhyaya on 2/10/16.
  */
class ReduceActor extends Actor {
  def receive: Receive = {
    case MapData(dataList) =>
      sender ! reduce(dataList)
  }
  def reduce(words: IndexedSeq[WordCount]): ReduceData = ReduceData {
    words.foldLeft(Map.empty[String, Int]) {  (index, words) =>
      if (index contains words.word)
        index + (words.word -> (index.get(words.word).get + 1))
      else
        index + (words.word -> 1)
    }
  }
}
